packages/runtime/container-runtime/src/test/gc/gcTelemetryTracker.spec.ts

```diff
@@ -0,0 +1,737 @@
+	function updateNodes(nodeIds: string[]) {
+		nodeIds.forEach((nodeId) => {
+			telemetryTracker.nodeUsed({
+				nodeId,
+				usageType: "Changed",
+				currentReferenceTimestampMs: Date.now(),
+				packagePath: testPkgPath,
+				completedGCRuns: 0,
+				isTombstoned: false,
+			});
+			telemetryTracker.nodeUsed({
+				nodeId,
+				usageType: "Loaded",
+				currentReferenceTimestampMs: Date.now(),
+				packagePath: testPkgPath,
+				completedGCRuns: 0,
+				isTombstoned: false,
+			});
+		});
+	}
+
+	// Mock node revived activity for the given nodes.
+	function reviveNode(fromId: string, toId: string, isTombstoned = false) {
+		telemetryTracker.nodeUsed({
+			nodeId: toId,
+			usageType: "Revived",
+			currentReferenceTimestampMs: Date.now(),
+			packagePath: testPkgPath,
+			completedGCRuns: 0,
+			isTombstoned,
+			fromId,
+		});
+		unreferencedNodesState.delete(toId);
+	}
+
+	/**
+	 * For summarizer clients, inactive / sweep ready events are not logged when on node usage. They are logged when GC
+	 * runs next time. This emulates that by calling the functions in the telemetry tracker that are called when GC runs.
+	 */
+	async function generateAllEvents(isSummarizerClient: boolean) {
+		if (!isSummarizerClient) {
+			return;
+		}
+		telemetryTracker.logSweepEvents(mc.logger, Date.now(), unreferencedNodesState, 0);
+		await telemetryTracker.logUnreferencedEvents(mc.logger);
+	}
+
+	before(() => {
+		clock = useFakeTimers();
+	});
+
+	beforeEach(() => {
+		mockLogger = new MockLogger();
+		mc = mixinMonitoringContext(
+			ChildLogger.create(mockLogger, "GarbageCollector"),
+			configProvider(injectedSettings),
+		);
+		unreferencedNodesState = new Map();
+	});
+
+	afterEach(() => {
+		clock.reset();
+		mockLogger.clear();
+		injectedSettings = {};
+	});
+
+	after(() => {
+		clock.restore();
+	});
+
+	// Tests that are run once for summarizer client and once for interactive client.
+	const clientTypeTests = (isSummarizerClient: boolean) => {
+		/**
+		 * Asserts that the events are as expected based on whether its a summarizer client or not. In non-summarizer
+		 * clients, only "InactiveObject_Loaded" and "SweepReadyObject_Loaded" events are logged. "Changed" and "Revived"
+		 * events are not logged.
+		 */
+		function assertMatchEvents(
+			events: Omit<ITelemetryBaseEvent, "category">[],
+			message: string,
+		) {
+			const expectedEvents: Omit<ITelemetryBaseEvent, "category">[] = [];
+			const unexpectedEvents: Omit<ITelemetryBaseEvent, "category">[] = [];
+			// For non-summarizer clients, events that are not "Loaded" are unexpected. Everything else is expected.
+			for (const event of events) {
+				const eventName = event.eventName as string;
+				if (!isSummarizerClient && !eventName.includes("Loaded")) {
+					unexpectedEvents.push(event);
+				} else {
+					expectedEvents.push(event);
+				}
+			}
+
+			// Note that mock logger clears all events after one of the `match` functions is called. Since we call match
```

Comment:
Yeah, if this use case gives you ideas for making mockLogger easier to use feel free to put a PR out.

Reply:
I thought about it but I could not find a generic way to do this. The only way I could think of is to either pass in a flag that says don't reset events or add a new function that doesn't reset events and both aren't great.

Reply:
Adding a function that asserts some events match and others don't might be a good option but I didn't want to add it for just this specific use case.